### ISSUE TITLE
tfjs.node.decodeImage run error in workthread

### ISSUE THREAD

**Issue by xnetsc** [tensorflow/tfjs#7064]
**System information**
any system，any node version 18.×

**Describe the current behavior**
in nodejs workThread:
```
ctx={tfjsnode:function(){return require("@tensorflow/tfjs-node")}};

code="imgData=fs.readFileSync("img.jpg");imgData=Uint8Array.from(imgData);tfjsnode().node.decodeImage(imgData)";

vm.runInNewContext(ctx,code);

```

the code would throw an error: decodeImage can only accept typedArray and so on.

i traced it into code and found:
isTypedArray(values) would always return false.
but really,values is a Uint8Array.

i don't know why.

**Describe the expected behavior**

**Standalone code to reproduce the issue**

**Other info / logs**

**Comment by mccoysc** [comment]
![image](https://user-images.githubusercontent.com/18025276/202887616-3400f2b1-ec54-4c18-b38d-c64af0f962a2.png)

**Comment by mccoysc** [comment]
i am sure that the ingData is "Uint8Array".
and the code can be run good in non-workThread and non-contextfied node env.

**Comment by rthadur** [comment]
You may need to use` imgData=Uint8Array(imgData) `instead of` imgData=Uint8Array.from(imgData)` , can you please check [here](https://github.com/tensorflow/tfjs/issues/4724) . Thank you!

**Comment by mccoysc** [comment]
maybe the offical should re-write the function code for "isTypedArray",the orignal code use instnace to determine if the content is an instance of "typed array".
but instance only checks if the object's constructor function is equal.

in fact,tensorflow has no need to be sure that the constructor functions are equal,it only needs making sure that object has the same actions.

**Comment by mattsoulanille** [comment]
I can reproduce this issue. Here's the code I'm using:
```js
const tfjs = require('@tensorflow/tfjs');
const tfjsnode = require('@tensorflow/tfjs-node');
const fs = require('fs');
const vm = require('node:vm');


console.log('Running in main context');
let imgData = fs.readFileSync('img.jpg');
imgData = Uint8Array.from(imgData);
console.log(tfjsnode.node.decodeImage(imgData));

const ctx = {
  tfjsnode: function() {
    return require('@tensorflow/tfjs-node');
  },
  fs: function() {
    return require('fs');
  },
  log: console.log,
};

code = `
log('Running in a new context');
let imgData = fs().readFileSync('img.jpg');
imgData = Uint8Array.from(imgData);
const decoded = tfjsnode().node.decodeImage(imgData);
log(decoded);
`;

vm.createContext(ctx);
vm.runInContext(code, ctx);
```

I'm not yet sure why this isn't working, but if you just need to load an image and run `decodeImage` on it, passing it directly instead of converting it to a Uint8Array seems to work in node 18:

```js
const vm = require('node:vm');

const ctx = {
  tfjsnode: function() {
    return require('@tensorflow/tfjs-node');
  },
  fs: function() {
    return require('fs');
  },
  log: console.log,
};

code = `
log('Running in a new context');
let imgData = fs().readFileSync('img.jpg');
// imgData = Uint8Array.from(imgData); // <-- This seems to break it.
const decoded = tfjsnode().node.decodeImage(imgData);
log(decoded);
`;

vm.createContext(ctx);
vm.runInContext(code, ctx);
```

**Comment by mattsoulanille** [comment]
> maybe the offical should re-write the function code for "isTypedArray",the orignal code use "instnaceof" to determine if the content is an instance of "typedarray". but "instanceof" only checks if the object's constructor function is equal.
> 
> in fact,"tensorflow" has no need to be sure that the constructor functions are equal. @rthadur

You're right, this seems to be the source of the issue. Thanks for the pointer!

Here's some test code that demonstrates it:
```js
const vm = require('node:vm');

const ctx = {
  tfjsnode: function() {
    return require('@tensorflow/tfjs-node');
  },
  fs: function() {
    return require('fs');
  },
  log: console.log,
};

code = `
log('Running in a new context');
let imgData = fs().readFileSync('img.jpg');
imgData = Uint8Array.from(imgData);
log('imgData instanceof Uint8Array:', imgData instanceof Uint8Array);
const tfn = tfjsnode();
log('tfn.util.isTypedArray(imgData):', tfn.util.isTypedArray(imgData));
`;

vm.runInNewContext(code, ctx);
```
This prints the following:
```
imgData instanceof Uint8Array: true
tfn.util.isTypedArray(imgData): false
```
This is unexpected because isTypedArray [returns true if its argument is an instance of Uint8Array](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/util_base.ts#L516-L520). Maybe there's something I don't understand about `vm.runInNewContext` that's making it check the `Uint8Array` of the image from the new context against the `Uint8Array` from the main context, and that's why it's failing?

**Comment by mattsoulanille** [comment]
My guess is that this code:
```js
const ctx = {
  tfjsnode: function() {
    return require('@tensorflow/tfjs-node');
  },
}
```
... is causing `tfjs-node` to be loaded in the main context while it should actually be loaded in the new context. That's why `isTypedArray` is returning false (since it's prototype is from the main context but it's checking an object from the new context). If we use the Uint8Array constructor from the main context as well, then it works:

```js
const fs = require('fs');
const vm = require('node:vm');

const ctx = {
  tfjsnode: function() {
    return require('@tensorflow/tfjs-node');
  },
  fs: function() {
    return require('fs');
  },
  log: console.log,
  makeUint8Array: function(data) {
    return new Uint8Array(data);
  },
};

code = `
log('Running in a new context');
let imgData = fs().readFileSync('img.jpg');
imgData = makeUint8Array(imgData);
log('imgData instanceof Uint8Array:', imgData instanceof Uint8Array);
const tfn = tfjsnode();
log('tfn.util.isTypedArray(imgData):', tfn.util.isTypedArray(imgData));
const decoded = tfn.node.decodeImage(imgData);
log(decoded);
`;

vm.runInNewContext(code, ctx);
```

Is there a way to load tfjs-node from the new context instead of the main context? I think that would solve this issue better than the above approach. Otherwise, I don't have a good solution to this. TFJS uses TypedArrays extensively to represent Tensor values, and we need to [check the type of a typed array](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/util.ts#L36-L40) when [creating a tensor](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/ops/tensor_ops_util.ts#L83-L86).

Maybe your problem can be solved with [WebWorkers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) in [node](https://nodejs.org/api/worker_threads.html) and [Comlink](https://github.com/GoogleChromeLabs/comlink) instead of `vm.runInNewContext`.

**Comment by mccoysc** [comment]
> My guess is that this code:
> 
> ```js
> 
> const ctx = {
> 
>   tfjsnode: function() {
> 
>     return require('@tensorflow/tfjs-node');
> 
>   },
> 
> }
> 
> ```
> 
> ... is causing `tfjs-node` to be loaded in the main context while it should actually be loaded in the new context. That's why `isTypedArray` is returning false (since it's prototype is from the main context but it's checking an object from the new context). If we use the Uint8Array constructor from the main context as well, then it works:
> 
> 
> 
> ```js
> 
> const fs = require('fs');
> 
> const vm = require('node:vm');
> 
> 
> 
> const ctx = {
> 
>   tfjsnode: function() {
> 
>     return require('@tensorflow/tfjs-node');
> 
>   },
> 
>   fs: function() {
> 
>     return require('fs');
> 
>   },
> 
>   log: console.log,
> 
>   makeUint8Array: function(data) {
> 
>     return new Uint8Array(data);
> 
>   },
> 
> };
> 
> 
> 
> code = `
> 
> log('Running in a new context');
> 
> let imgData = fs().readFileSync('img.jpg');
> 
> imgData = makeUint8Array(imgData);
> 
> log('imgData instanceof Uint8Array:', imgData instanceof Uint8Array);
> 
> const tfn = tfjsnode();
> 
> log('tfn.util.isTypedArray(imgData):', tfn.util.isTypedArray(imgData));
> 
> const decoded = tfn.node.decodeImage(imgData);
> 
> log(decoded);
> 
> `;
> 
> 
> 
> vm.runInNewContext(code, ctx);
> 
> ```
> 
> 
> 
> Is there a way to load tfjs-node from the new context instead of the main context? I think that would solve this issue better than the above approach. Otherwise, I don't have a good solution to this. TFJS uses TypedArrays extensively to represent Tensor values, and we need to [check the type of a typed array](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/util.ts#L36-L40) when [creating a tensor](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/ops/tensor_ops_util.ts#L83-L86).
> 
> 
> 
> Maybe your problem can be solved with [WebWorkers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) in [node](https://nodejs.org/api/worker_threads.html) and [Comlink](https://github.com/GoogleChromeLabs/comlink) instead of `vm.runInNewContext`.

as you said,tfjs only use "Uint8Array" to "represent" tensor data,no need to care about if it's really a "Uint8Array".

so, "instanceof" is no need.

**Comment by mattsoulanille** [comment]
Whether this is possible or not, I don't think it does what you want. When you run `tfjsnode()` in the new context to get access to tfjs-node, node still loads tfjs-node in the main context. Even though you use `vm.runInNewContext`, this `tfjsnode()` code is not isolated because `require` loads it in the main context. To get proper isolation, you may need to reimplement `require` in the new context (maybe by reading and `eval`-ing the file or something equivalent), but I don't recommend this.

Can you share the problem you're solving with `vm.runInNewContext`? What is the use-case for doing this? Thanks!

**Comment by mccoysc** [comment]
> Whether this is possible or not, I don't think it does what you want. When you run `tfjsnode()` in the new context to get access to tfjs-node, node still loads tfjs-node in the main context. Even though you use `vm.runInNewContext`, this `tfjsnode()` code is not isolated because `require` loads it in the main context. To get proper isolation, you may need to reimplement `require` in the new context (maybe by reading and `eval`-ing the file or something equivalent), but I don't recommend this.
> 
> 
> 
> Can you share the problem you're solving with `vm.runInNewContext`? What is the use-case for doing this? Thanks!

no.
default,no "require" can be called in new context in nodejs.
if you want,the only way to "require" is give a "require" function in context params.

so,as my code,it's the only way.

**Comment by mccoysc** [comment]
the code running in new context is uploaded by user and must be isolated by new context.&nbsp;


and,by the definetion of nodejs for vm context:
in new context,you can only use the v8 basic interface,but no any node module can be used,require ,fs for example.
if you want,you need inject the main thread module or interface to new context.



发自我的iPhone


------------------ Original ------------------
From: Matthew Soulanille ***@***.***&gt;
Date: Thu,Nov 24,2022 1:55 AM
To: tensorflow/tfjs ***@***.***&gt;
Cc: mccoysc ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [tensorflow/tfjs] tfjs.node.decodeImage run error in workthread (Issue #7064)






Whether this is possible or not, I don't think it does what you want. When you run tfjsnode() in the new context to get access to tfjs-node, node still loads tfjs-node in the main context. Even though you use vm.runInNewContext, this tfjsnode() code is not isolated because require loads it in the main context. To get proper isolation, you may need to reimplement require in the new context (maybe by reading and eval-ing the file or something equivalent), but I don't recommend this.

Can you share the problem you're solving with vm.runInNewContext? What is the use-case for doing this? Thanks!

**Comment by mattsoulanille** [comment]
I will be out of office Thursday and Friday, but I'm going to reopen this so I can take another look at it next week.

**Comment by RaisinTen** [comment]
> the code running in new context is uploaded by user and must be isolated by new context.

Is this for security? The Node.js docs for VM - https://nodejs.org/api/vm.html#vm-executing-javascript already says:

> **The `node:vm` module is not a security mechanism. Do not use it to run untrusted code.**

**Comment by mccoysc** [comment]
> > the code running in new context is uploaded by user and must be isolated by new context.
> 
> 
> 
> Is this for security? The Node.js docs for VM - https://nodejs.org/api/vm.html#vm-executing-javascript already says:
> 
> 
> 
> > **The `node:vm` module is not a security mechanism. Do not use it to run untrusted code.**

no,the "user code" are all audited by some trusted org.
i isolate some runtime vars by vm context.

**Comment by RaisinTen** [comment]
What do you think about copying everything except for your context specific runtime vars for each vm script?

I modified you example and now it returns true in both cases FWIW:
```js
const vm = require('node:vm');

const ctx = vm.createContext(Object.defineProperties(
  {
    ...global,
    tfjsnode: function() {
      return require('@tensorflow/tfjs-node');
    },
    log: console.log,
  },
  Object.getOwnPropertyDescriptors(global),
));

code = `
log('Running in a new context');
const imgData = Uint8Array.from([1,2]);
log('imgData instanceof Uint8Array:', imgData instanceof Uint8Array);
const tfn = tfjsnode();
log('tfn.util.isTypedArray(imgData):', tfn.util.isTypedArray(imgData));
`;

vm.runInNewContext(code, ctx);
```

**Comment by mccoysc** [comment]
> What do you think about copying everything except for your context specific runtime vars for each vm script?
> 
> I modified you example and now it returns true in both cases FWIW:

not good,i need isolate all vars to make user code run in new env every time.

**Comment by RaisinTen** [comment]
> not good,i need isolate all vars to make user code run in new env every time.

I mean all the variables are isolated in the current approach, that's why Uint8Array in the main context is different from Uint8Array in the vm script's context. It looks like you want the variables on different contexts to be the same and different at the same time?

If you want everything to be isolated, it doesn't make sense to share the contents of a module you required in the main context with the vm script's context.

If you are okay with using the variables of the main context and the vm script's context, you should use `vm.runInThisContext`, where everything would run in the same context.

These can't be mixed together.

You might find the discussion in https://github.com/nodejs/help/issues/761#issuecomment-318865290 useful.

**Comment by mccoysc** [comment]
yes,i want isolate all vars in new context,not in "this context".
yes,better,if i can require module in new context, but can't.

**Comment by RaisinTen** [comment]
> yes,i want isolate all vars in new context,not in "this context".
> yes,better,if i can require module in new context, but can't.

Yea, I don't see a way of achieving what you want without implementing your own `require()` in the vm script's context without using anything from the main context.

**Comment by mccoysc** [comment]
maybe node should add "default" require to vm context and also can be overwrite by user context?





------------------ Original ------------------
From: Darshan Sen ***@***.***&gt;
Date: Fri,Nov 25,2022 11:39 PM
To: tensorflow/tfjs ***@***.***&gt;
Cc: mccoysc ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [tensorflow/tfjs] tfjs.node.decodeImage run error in workthread (Issue #7064)






yes,i want isolate all vars in new context,not in "this context".
yes,better,if i can require module in new context, but can't.

Yea, I don't see a way of achieving what you want without implementing your own require() in the vm script's context without using anything from the main context.

**Comment by RaisinTen** [comment]
There is an open issue in https://github.com/nodejs/node/issues/31852 that pretty much asks for something like that.

I think the easiest way to unblock yourself would be to pass Uint8Array to the vm script's context. I know that it's a hack but it would eventually go away once that issue gets fixed and looking at some of the recent comments it looks like we are getting close.

**Comment by Varun-S10** [comment]
Hi @xnetsc, Thank you for bringing this to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by Varun-S10** [comment]
Hi @xnetsc, Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it. Thank you for your understanding.
